DSCResources/MSFT_xAdcsCertificationAuthority/MSFT_xAdcsCertificationAuthority.psm1

```diff
@@ -1,162 +1,485 @@
 {
     [OutputType([System.Collections.Hashtable])]
     [CmdletBinding()]
-    param(
-    [ValidateSet('Present','Absent')]
-    [string]$Ensure = 'Present',
-    [string]$CACommonName,
-    [string]$CADistinguishedNameSuffix,
-    [Parameter(Mandatory)]
-    [ValidateSet('EnterpriseRootCA','EnterpriseSubordinateCA','StandaloneRootCA','StandaloneSubordinateCA')]
-    [string]$CAType,
-    [string]$CertFile,
-    [pscredential]$CertFilePassword,
-    [string]$CertificateID,
-    [Parameter(Mandatory)]
-    [pscredential]$Credential,
-    [string]$CryptoProviderName,
-    [string]$DatabaseDirectory,
-    [string]$HashAlgorithmName,
-    [boolean]$IgnoreUnicode,
-    [string]$KeyContainerName,
-    [uint32]$KeyLength,
-    [string]$LogDirectory,
-    [string]$OutputCertRequestFile,
-    [boolean]$OverwriteExistingCAinDS,
-    [boolean]$OverwriteExistingDatabase,
-    [boolean]$OverwriteExistingKey,
-    [string]$ParentCA,
-    [ValidateSet('Hours','Days','Months','Years')]
-    [string]$ValidityPeriod,
-    [uint32]$ValidityPeriodUnits
+    param (
+        [Parameter(Mandatory)]
+        [ValidateSet('EnterpriseRootCA','EnterpriseSubordinateCA','StandaloneRootCA','StandaloneSubordinateCA')]
+        [string] $CAType,
+
+        [Parameter(Mandatory)]
+        [pscredential] $Credential,
+
+        [ValidateSet('Present','Absent')]
+        [string] $Ensure = 'Present',
+
+        [string] $CACommonName,
+
+        [string] $CADistinguishedNameSuffix,
+
+        [string] $CertFile,
+
+        [pscredential] $CertFilePassword,
+
+        [string] $CertificateID,
+
+        [string] $CryptoProviderName,
+
+        [string] $DatabaseDirectory,
+
+        [string] $HashAlgorithmName,
+
+        [boolean] $IgnoreUnicode,
+
+        [string] $KeyContainerName,
+
+        [uint32] $KeyLength,
+
+        [string] $LogDirectory,
+
+        [string] $OutputCertRequestFile,
+
+        [boolean] $OverwriteExistingCAinDS,
+
+        [boolean] $OverwriteExistingDatabase,
+
+        [boolean] $OverwriteExistingKey,
+
+        [string] $ParentCA,
+
+        [ValidateSet('Hours','Days','Months','Years')]
+        [string] $ValidityPeriod,
+
+        [uint32] $ValidityPeriodUnits
     )
 
+    Write-Verbose -Message ( @(
+            "$($MyInvocation.MyCommand): "
+            $($LocalizedData.GettingAdcsCAStatusMessage -f $CAType)
+        ) -join '' )
+
     $ADCSParams = @{} + $PSBoundParameters
-    $ADCSParams.Remove('Ensure') | out-null
-    $ADCSParams.Remove('Debug') | out-null
-    $ADCSParams.Remove('ErrorAction') | out-null
+    $null = $ADCSParams.Remove('Ensure')
+    $null = $ADCSParams.Remove('Debug')
+    $null = $ADCSParams.Remove('ErrorAction')
+
+    try
+    {
+        $null = Install-AdcsCertificationAuthority @ADCSParams -WhatIf
+        # CA is not installed
+        $Ensure = 'Absent'
+    }
+    catch
+    {
+        # CA is probably already installed - output the exception just in case
+        # it is some other problem.
```

Comment:
If it is  possible to know the exact exception that we will get when CA is already installed, is it not better to look for that explicitly and use that as the criteria for this case?